### DISCUSSION TITLE
Question: `txpool` behaviour after dropping unpayable tx

### DISCUSSION THREAD

**Discussion by kamikazechaser** [celo-org/celo-blockchain#2050] [Help (Q&A)]
To reproduce:

I manage my own nonces and push a lot of transactions fast to the node. In some cases, if my balance is not enough, the node will accept the tx but then immediately drop it. Any future tx that I send gets rejected. If I top up my gas balance and try to resubmit the tx, it goes into a pending state (txpool_inspect) but never gets mined. Restarting the node will finally get it mined. Here is a timeline roughly describing the scenario:

```sh
# My balance here = 1
tx A with nonce 14 and gas cost 0.3 - accepted and mined
tx B with nonce 15 and gas cost 0.8 - accepted
tx B dropped (logged as TRACE: Removed unpayable pending transaction)
# tx B is not visible with txpool_inspect
tx C with nonce 16 and gas cost 0.8 - rejected (not enough gas)
# My balance here = 0.7
# I top up 1 more CELO, new balance 1.7
tx D is a resubmit of tx B unmodified i.e. with same nonce 15 and gas cost 0.8 - accepted
# tx D is now visible in txpool_inspect as pending but doesn't get mined
# It gets accepted by other nodes but won't get mined
# tx D only gets mined after restarting the node
```

My question is:

Why does the resubmitted tx go into an indefinite pending state yet it seemingly seems to fulfill txpool accept conditions (enough gas, correct next nonce)? Whereas restarting the local node (which will load the same pending tx from the journal) will get it mined immediately.

Is there still a reference to tx B somehow even after it is dropped? tx C leads me to believe so.

**Comment by palango** [comment]
@kamikazechaser Do you see any trace messages after tx `C` and `D` are submitted?

**Comment by kamikazechaser** [comment]
I did not save the trace dump but will get a new one asap.
